### DISCUSSION TITLE
TensorRT failing to export

### DISCUSSION THREAD

**Discussion by ajrenzo** [ultralytics/ultralytics#1946] [Q&A]
I'm trying to export some models into .engine format to increase their inference speed, but I they keep failing despite the onnx feature completing successfully. I've even tried it on the yolov8n.pt model and still no luck. The errors seem to be the same: 

[04/10/2023-21:41:52] [TRT] [E] ModelImporter.cpp:774: ERROR: builtin_op_importers.cpp:679 In function importConv:
[6] Assertion failed: (shapedBiasWeights.shape.nbDims == 1) && "The bias tensor is required to be 1D."

I'm using python 3.10, TensorRT-8.6.0.12 with CUDA 12.0. The code is as follows: 

```
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
exportedmodel = model.export( format = 'engine', device = 0)
```

The output for the above is as follows:

```
Ultralytics YOLOv8.0.59  Python-3.10.0 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 4090, 24564MiB)
YOLOv8n summary (fused): 168 layers, 3151904 parameters, 0 gradients, 8.7 GFLOPs

PyTorch: starting from yolov8n.pt with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (6.2 
MB)

ONNX: starting export with onnx 1.13.1...
ONNX: simplifying with onnxsim 0.4.19...
ONNX: export success  1.1s, saved as yolov8n.onnx (0.1 MB)

TensorRT: starting export with TensorRT 8.6.0...
[04/10/2023-21:41:46] [TRT] [I] [MemUsageChange] Init CUDA: CPU +323, GPU +0, now: CPU 50113, GPU 5382 (MiB)
[04/10/2023-21:41:52] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +1424, GPU +539, now: CPU 52702, GPU 5921 (MiB)
[04/10/2023-21:41:52] [TRT] [W] CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage and speed up TensorRT initialization. See "Lazy Loading" section of CUDA documentation https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#lazy-loading
[04/10/2023-21:41:52] [TRT] [I] ----------------------------------------------------------------
[04/10/2023-21:41:52] [TRT] [I] Input filename:   yolov8n.onnx
[04/10/2023-21:41:52] [TRT] [I] ONNX IR version:  0.0.8
[04/10/2023-21:41:52] [TRT] [I] Opset version:    17
[04/10/2023-21:41:52] [TRT] [I] Producer name:    pytorch
[04/10/2023-21:41:52] [TRT] [I] Producer version: 1.13.1
[04/10/2023-21:41:52] [TRT] [I] Domain:
[04/10/2023-21:41:52] [TRT] [I] Model version:    0
[04/10/2023-21:41:52] [TRT] [I] Doc string:
[04/10/2023-21:41:52] [TRT] [I] ----------------------------------------------------------------
[04/10/2023-21:41:52] [TRT] [E] ModelImporter.cpp:768: While parsing node number 0 [Conv -> "/model.0/conv/Conv_output_0"]:
[04/10/2023-21:41:52] [TRT] [E] ModelImporter.cpp:769: --- Begin node ---
[04/10/2023-21:41:52] [TRT] [E] ModelImporter.cpp:770: input: "images"
input: "model.0.conv.weight"
input: "model.0.conv.weight"
output: "/model.0/conv/Conv_output_0"
name: "/model.0/conv/Conv"
op_type: "Conv"
attribute {
  name: "dilations"
  ints: 1
  ints: 1
  type: INTS
}
attribute {
  name: "group"
  i: 1
  type: INT
}
attribute {
  name: "kernel_shape"
  ints: 3
  ints: 3
  type: INTS
}
attribute {
  name: "pads"
  ints: 1
  ints: 1
  ints: 1
  ints: 1
  type: INTS
}
attribute {
  name: "strides"
  ints: 2
  ints: 2
  type: INTS
}

[04/10/2023-21:41:52] [TRT] [E] ModelImporter.cpp:771: --- End node ---
[04/10/2023-21:41:52] [TRT] [E] ModelImporter.cpp:774: ERROR: builtin_op_importers.cpp:679 In function importConv:
[6] Assertion failed: (shapedBiasWeights.shape.nbDims == 1) && "The bias tensor is required to be 1D."
TensorRT: export failure  7.9s: failed to load ONNX file: yolov8n.onnx
```

I'm wondering if this is because TensorRT 8.6 is so new (something similar happened with opencv), but this is the only version that seems to work with CUDA v12.0

**Comment by glenn-jocher** [comment]
@ajrenzo TensorRT export works correctly. I'd recommend exporting in a verified environment like Docker if you are having environment issues. 


## Environments

YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

**Reply by ajrenzo** [reply]
Thanks! And thank goodness for Google Colab (I run on windows so Docker may not work). Now to figure out why Visual Studio Code cant do the same thing... Yolov5 seems to export just fine, although it runs into a torch.Size not equal to max model size issue. I'm wondering if it might be an issue with CUDNN

**Reply by ajrenzo** [reply]
While Google Colab seems to work in exporting the file, I'm still unable to run it on my machine as it claims to require the same version of tensorrt. I noticed Google Colab uses a linux version of tensorrt, but I have to run my code on Windows (in Visual Studio Code, although I've also run on Visual Studio). I'm still getting the same error as above regardless of what I do or what I replace. Is this perhaps a Windows issue?

**Reply by glenn-jocher** [reply]
@ajrenzo it's great to hear that Google Colab worked for you. The issue you're encountering on your Windows machine could indeed be related to the Windows environment or a compatibility issue between TensorRT and CUDA versions. Windows support for these tools can sometimes lag behind Linux, or there may be subtle differences in implementation that cause issues like the one you're experiencing.

If you're consistently facing this problem on Windows, it might be worth considering the following steps:

1. **Verify Compatibility**: Ensure that the versions of CUDA, cuDNN, and TensorRT you're using are compatible with each other and with your PyTorch version. NVIDIA typically provides compatibility matrices that can help with this.

2. **Update Drivers**: Make sure your NVIDIA drivers are up to date, as outdated drivers can sometimes cause issues with CUDA and TensorRT functionality.

3. **Environment Consistency**: Try to replicate the environment settings from Colab as closely as possible on your Windows machine, including the versions of all libraries and tools.

4. **TensorRT on Windows**: There might be specific instructions or additional steps required to properly set up TensorRT on Windows. Check the official TensorRT documentation for any Windows-specific installation guidelines.

5. **Alternative Approaches**: If possible, consider using a Linux virtual machine or a dual-boot setup on your Windows machine to create a Linux environment that might be more compatible with the tools you're using.

6. **Community and Support**: Engage with the community on forums or the official GitHub repository issues page. Other users may have encountered similar problems and could have solutions or workarounds.

7. **Ultralytics Support**: If you have an enterprise license, you might be eligible for direct support from Ultralytics, which could provide more targeted assistance for your issue.

Remember that while YOLOv5 and YOLOv8 share similarities, they are different codebases and may have different dependencies or compatibility considerations. Keep troubleshooting systematically, and you'll likely find a solution that works for your setup.

**Comment by ajrenzo** [comment]
Looks like this may be a Windows issue with YoloV8 (could someone confirm?). After setting up WSL I was able to get .engine file to successfully export, although the .engine file still wouldnt work when I ported it over to Windows; instead, the program just stalled. I cant use WSL as it seems I cant take screenshot data from the Windows side of things, and some of the modules I use require Windows.

Currently running CUDA 12.1 with CUDNN 8.9 and TensorRT EA 8.6; All of this is being run on Windows 10 in Visual Studio Code. The same modules are running on WSL Ubuntu 22.04.

YoloV5 does seem to export successfully, although with custom models it returns an error with the torch.Size not being equal to the max model size despite training and exporting it at 640pixels. Not sure if this is related but figured the more info the better.

If someone's managed to get TensorRT exported and running with YoloV8 on Windows, is there anything you did to achieve this?

**Reply by huyvt811** [reply]
I face same issue like you mentioned above and I don't know how to resolve this

**Reply by huyvt811** [reply]
I already fix this issue. If you need help, contact me

**Reply by ajrenzo** [reply]
how did you fix it?

**Reply by huyvt811** [reply]
Hi @ajrenzo , I missed the notification so now I responds to you late. Hope this help, if you still has problem, keep responds it.

First, you need to downgrade TensorRT-8.6.0.12 to TensorRT-8.4.3.1. I remember that when I downgrade to TensorRT-8.5.x.x, log will return,
`[6] Assertion failed: (splitList.empty() || (splitList.size() == numOutputs)) && "The shape of the split attribute misaligns with the number of outputs."`
So I think you need to fix version TensorRT at 8.4.3.1. This version is being used in Google Colab environment.

After that, you need to update onnnx_sim, with version >= 0.4.21.
`ONNX: simplifying with onnxsim 0.4.19... `
your version is 0.4.19 which I see in your log you post.
